### PR TITLE
satellite/metabase/rangedloop: continue after error

### PR DESCRIPTION
**Pull request by Erikvv**
When an observer errors we still want to finish the other observers.

This changes store the error and continues the loop, skipping
the observer which errored out and setting the duration metric to -1.

When the error occurs in the process stage, it does continue the other
ranges of the same observer. It removes the observer entirely after the process
stage. To improve this would make it more complex due to race
conditions.
 
Please describe the performance impact: insignificant

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/main/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/main/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/metabase/rangedloop/rangedlooptest/infiniteprovider.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ func (m *InfiniteSegmentProvider) CreateRanges(nRanges int, batchSize int) (segm
 func (m *InfiniteSegmentProvider) Iterate(ctx context.Context, fn func([]segmentloop.Segment) error) error {
 	for {
 		err := fn(make([]segmentloop.Segment, 3))
+		println("InfiniteSegmentProvider.Iterate")
```
Comment (mniewrzal) [inline satellite/metabase/rangedloop/rangedlooptest/infiniteprovider.go:32]: do we need it?

File: satellite/metabase/rangedloop/service_test.go [hunk 2]
```diff
@@ -149,3 +151,202 @@ func TestLoopCancellation(t *testing.T) {
...
+			Segments: segments,
+		},
+		observers,
+	)
+
+	observerDurations, err := loopService.RunOnce(testcontext.New(t))
+	require.NoError(t, err)
+	// only the 2 completed observers should have the duration reported
+	require.Len(t, observerDurations, len(observers))
+	require.Equal(t, int32(38), shouldBeCalled)
```
Comment (mniewrzal) [inline satellite/metabase/rangedloop/service_test.go:340]: nit: use EqualValues

also wondering how we can make 38 less magical

File: satellite/metabase/rangedloop/service.go [hunk 3]
```diff
@@ -195,30 +201,79 @@ func finishObservers(ctx context.Context, observerStates []observerState) (obser
 
 // Iterating over the segments is done.
 // This is the reduce step.
-func finishObserver(ctx context.Context, state observerState) (ObserverDuration, error) {
+func finishObserver(ctx context.Context, log *zap.Logger, state observerState) ObserverDuration {
+	if state.err != nil {
+		return ObserverDuration{
+			Observer: state.observer,
+			Duration: -1 * time.Second,
```
Comment (mniewrzal) [inline satellite/metabase/rangedloop/service.go:208]: won't 0 have the same effect here (and in other places)? that observer have no duration?

File: satellite/metabase/rangedloop/service.go [hunk 4]
```diff
@@ -52,20 +53,25 @@ func NewService(log *zap.Logger, config Config, provider RangeSplitter, observer
...
 	rangeObservers []*rangeObserverState
+	// err is the error that occurred during the observer's Start method.
+	// If err is set, the observer will be skipped during the loop iteration.
+	err error
 }
 
 type rangeObserverState struct {
 	rangeObserver Partial
 	duration      time.Duration
+	err           error
```
Comment (mniewrzal) [inline satellite/metabase/rangedloop/service.go:67]: should we add also similar command like for observerState errro? something like `If err is set, the range observer will be skipped during the loop iteration.`

File: satellite/metabase/rangedloop/service.go (hunk 3 above)
Comment (Erikvv) [reply satellite/metabase/rangedloop/service.go:208]: I think -1 seems more explicitly an error than 0, which could be a valid duration.

File: satellite/metabase/rangedloop/service_test.go (hunk 2 above)
Comment (Erikvv) [reply satellite/metabase/rangedloop/service_test.go:340]: I will split this out into 5 counters:

- numOnStartCalls
- numOnForkCalls
- numOnProcessCalls
- numOnJoinCalls
- numOnFinishCalls

**Review by mniewrzal — APPROVED** [review]

File: satellite/metabase/rangedloop/service.go [hunk 5]
```diff
@@ -52,20 +53,27 @@ func NewService(log *zap.Logger, config Config, provider RangeSplitter, observer
...
 	duration      time.Duration
+	// err is the error that is returned by the observer's Fork or Process method.
+	// If err is set, the range observer will be skipped during the loop iteration.
+	err error
 }
 
 // ObserverDuration reports back on how long it took the observer to process all the segments.
 type ObserverDuration struct {
 	Observer Observer
+	// When this observer errored out, set duration to -1s
```
Comment (azdagron) [inline satellite/metabase/rangedloop/service.go:75]: nitpick: comment should start with the field name, i.e. `Duration`

**Review by azdagron — APPROVED** [review]

File: satellite/metabase/rangedloop/service.go (hunk 5 above)
Comment (Erikvv) [reply satellite/metabase/rangedloop/service.go:75]: I'll sneak that in another change so I don't lose the approve.

Comment (Erikvv) [reply satellite/metabase/rangedloop/service.go:75]: Done in https://review.dev.storj.io/c/storj/storj/+/9156

**Pull request merged**